tests/testlib/s2n_ktls_test_utils.c

```diff
@@ -0,0 +1,185 @@
+
+    size_t total_len = 0;
+    for (size_t count = 0; count < msg->msg_iovlen; count++) {
+        uint8_t *buf = msg->msg_iov[count].iov_base;
+        POSIX_ENSURE_REF(buf);
+        size_t len = msg->msg_iov[count].iov_len;
+
+        if (s2n_stuffer_write_bytes(&io_ctx->data_buffer, buf, len) < 0) {
+            /* This mock implementation doesn't handle partial writes for msg_iovlen > 1.
+             *
+             * This simplifies the implementation and importantly doesn't limit our test
+             * coverage because partial write are handled the same regardless of
+             * msg_iovlen. */
+            POSIX_ENSURE(count == 0, S2N_ERR_SAFETY);
+
+            errno = EAGAIN;
+            return -1;
+        }
+
+        total_len += len;
+    }
+    if (total_len) {
+        /* write record_type and len after some data was written successfully */
+        POSIX_GUARD(s2n_stuffer_write_uint8(&io_ctx->ancillary_buffer, *record_type));
+        POSIX_GUARD(s2n_stuffer_write_uint16(&io_ctx->ancillary_buffer, total_len));
+    }
+
+    return total_len;
+}
+
+/* In userspace TLS, s2n first reads the header to determine the length of next record
+ * and then reads the entire record into conn->in. In kTLS it is not possible to know
+ * the length of the next record. Instead the socket returns the minimum of
+ * bytes-requested and data-available; reading multiple consecutive records if they
+ * are of the same type. */
+ssize_t s2n_test_ktls_recvmsg_io_stuffer(void *io_context, struct msghdr *msg)
+{
+    POSIX_ENSURE_REF(io_context);
+    POSIX_ENSURE_REF(msg);
+    POSIX_ENSURE_REF(msg->msg_iov);
+
+    /* Assuming msg_control is uint8_t is a simplification and will not work when we
+     * attempt to test the production s2n_ktls_recv implementation. However, setting/parsing
+     * cmsg is critical code and will be added in a separate PR. */
+    uint8_t *record_type = (uint8_t *) msg->msg_control;
+    POSIX_ENSURE_REF(record_type);
+    struct s2n_test_ktls_io_stuffer *io_ctx = (struct s2n_test_ktls_io_stuffer *) io_context;
+    POSIX_ENSURE_REF(io_ctx);
+    io_ctx->send_recv_msg_invoked_count++;
+    /* s2n only receives using msg_iovlen of 1 */
+    POSIX_ENSURE_EQ(msg->msg_iovlen, 1);
+    uint8_t *buf = msg->msg_iov->iov_base;
+    POSIX_ENSURE_REF(buf);
+
+    /* There is no data available so return blocked */
+    if (!s2n_stuffer_data_available(&io_ctx->ancillary_buffer)) {
+        errno = EAGAIN;
+        return -1;
+    }
+
+    ssize_t total_read = 0;
+    size_t amount_requested = msg->msg_iov->iov_len;
+    while (true) {
+        /* read record_type and number of bytes available in the next record */
+        POSIX_GUARD(s2n_stuffer_read_uint8(&io_ctx->ancillary_buffer, record_type));
+        uint16_t n_avail = 0;
+        POSIX_GUARD(s2n_stuffer_read_uint16(&io_ctx->ancillary_buffer, &n_avail));
+        POSIX_ENSURE_LTE(n_avail, s2n_stuffer_data_available(&io_ctx->data_buffer));
+
```

Comment:
This failure will actually be more generic, since it'll just be S2N_ERR_SAFETY. And I don't think the location of the error is any better-- here vs where you actually try to read.